cpp/src/arrow/c/abi.h

```diff
@@ -228,6 +228,65 @@ struct ArrowDeviceArrayStream {
 
 #endif  // ARROW_C_DEVICE_STREAM_INTERFACE
 
+#ifndef ARROW_C_ASYNC_STREAM_INTERFACE
+#define ARROW_C_ASYNC_STREAM_INTERFACE
+
+// Similar to ArrowDeviceArrayStream, except designed for an asynchronous
+// style of interaction. While ArrowDeviceArrayStream provides producer
+// defined callbacks, this is intended to be created by the consumer instead.
+// The consumer passes this handler to the producer, which in turn uses the
+// callbacks to inform the consumer of events in the stream.
+struct ArrowAsyncDeviceStreamHandler {
+  // Handler for receiving a schema. The passed in stream_schema should be
+  // released or moved by the handler (producer is giving ownership of it to
+  // the handler).
+  //
+  // The `extension_param` argument can be null or can be used by a producer
```

Comment:
How should it be freed/what is the lifetime + ownership? (Especially in the case where the consumer may not understand the format of the extension and won't know how to free it?)

Reply:
I read this as "the contents of `stream_schema` are only valid during the call.  Copy off any information that you need.  The `ArrowSchema` will be deleted by the producer at some point after the call completes."  Or, to put it another way "the consumer shall not call `release` on the `ArrowSchema`"

Reply:
I guess I should reword it a bit then, the intent was that the ownership of the `ArrowSchema` is being given to the consumer from the Producer, thus the consumer *MUST* call `release` on the `ArrowSchema` unless it needs to keep it.

Reply:
> How should it be freed/what is the lifetime + ownership? (Especially in the case where the consumer may not understand the format of the extension and won't know how to free it?)

The intent was that the `extension_param` is entirely owned by the producer, the consumer shouldn't care or need to manage the lifetime of the extension at all.

Reply:
Maybe it could be `const void*` then?

Reply:
The fact that it's `void*` implies that there's out-of-band communication between the producer and consumer. For example:

- the consumer is a GUI and the producer is a dataset library
- the consumer initiates an interaction by passing
  - a query string
  - an `ArrowAsyncDeviceStreamHandler`
  - `ExtensionParamKind::TOTAL_NUMBER_OF_ROWS`

Since the producer received that last argument, it will ensure that `extension_param` points to an `int64_t` (or fail, or leave it null to indicate that capability is not available, or ...)

Since this scenario already assumes the producer and consumer can communicate what the extension parameter should be without recourse to the arrow C ABI, it seems odd to give them another more restrictive channel to communicate non-arrow objects.

I think this parameter should either be removed as out of scope or replaced with something accessible to consumers without out-of-band typeinfo (as @paleolimbot suggested, [metadata](https://arrow.apache.org/docs/format/CDataInterface.html#c.ArrowSchema.metadata) would work)